File: tools/mfo-dashboard/mcp_client.py
"""MCP JSON-RPC клиент для insapp-db"""

import json
import threading
from concurrent.futures import ThreadPoolExecutor
import requests as http_requests
from config import MCP_URL, MCP_KEY


class MCPClient:
    def __init__(self):
        self.session_id = None
        self.req_id = 0
        self._lock = threading.Lock()
        self._http = http_requests.Session()

    def _next_id(self):
        with self._lock:
            self.req_id += 1
            return self.req_id

    def _headers(self):
        h = {
            "Content-Type": "application/json",
            "x-api-key": MCP_KEY,
            "Accept": "application/json, text/event-stream",
        }
        if self.session_id:
            h["mcp-session-id"] = self.session_id
        return h

    def _ensure_session(self):
        if self.session_id:
            return
        try:
            resp = self._http.post(MCP_URL, json={
                "jsonrpc": "2.0", "method": "initialize",
                "params": {
                    "protocolVersion": "2024-11-05",
                    "capabilities": {},
                    "clientInfo": {"name": "mfo-dashboard", "version": "1.0"},
                },
                "id": self._next_id(),
            }, headers=self._headers(), timeout=10)
            self.session_id = resp.headers.get("mcp-session-id") or resp.headers.get("Mcp-Session")
            self._http.post(MCP_URL, json={
                "jsonrpc": "2.0", "method": "notifications/initialized", "params": {},
            }, headers=self._headers(), timeout=5)
            print(f"MCP session: {self.session_id}")
        except Exception as e:
            print(f"MCP init error: {e}")

    def query_parallel(self, queries):
        """Run multiple queries in parallel. queries = [(sql, desc), ...]. Returns list of row-lists."""
        self._ensure_session()
        with ThreadPoolExecutor(max_workers=len(queries)) as pool:
            return list(pool.map(lambda q: self.query(q[0], q[1]), queries))

    def query(self, sql, desc="dashboard"):
        self._ensure_session()
        for attempt in range(3):
            try:
                resp = self._http.post(MCP_URL, json={
                    "jsonrpc": "2.0", "method": "tools/call",
                    "params": {"name": "query", "arguments": {
                        "database": "InsappCoreProd",
                        "sql": sql,
                        "user_prompt": "dashboard",
                        "query_description": desc,
                    }},
                    "id": self._next_id(),
                }, headers=self._headers(), timeout=120)

                ct = resp.headers.get("Content-Type", "")
                if "text/event-stream" in ct:
                    for line in resp.text.split("\n"):
                        if line.startswith("data: "):
                            return self._parse(json.loads(line[6:]))
                else:
                    return self._parse(resp.json())
            except (http_requests.exceptions.Timeout, http_requests.exceptions.ConnectionError):
                print(f"MCP timeout/conn error (attempt {attempt+1}/3): {desc}")
                if attempt < 2:
                    continue
            except Exception as e:
                print(f"MCP query error: {e}")
                break
        return []

    def _parse(self, data):
        try:
            for item in data.get("result", {}).get("content", []):
                if item.get("type") == "text":
                    parsed = json.loads(item["text"])
                    if parsed.get("success"):
                        return parsed.get("rows", [])
                    if parsed.get("error"):
                        print(f"SQL error: {parsed['error']}")
        except Exception as e:
            print(f"Parse error: {e}, data: {str(data)[:200]}")
        return []

File: tools/mfo-dashboard/Services/McpClient.cs
using System.Net;
using System.Text;
using System.Text.Json;
using DashboardApi.Models;
using Microsoft.Extensions.Options;

namespace DashboardApi.Services;

public class McpClient
{
    private readonly IHttpClientFactory _httpFactory;
    private readonly McpSettings _settings;
    private volatile string? _sessionId;
    private int _reqId;
    private readonly SemaphoreSlim _sessionLock = new(1, 1);

    public McpClient(IHttpClientFactory httpFactory, IOptions<McpSettings> settings)
    {
        _httpFactory = httpFactory;
        _settings = settings.Value;
    }

    private int NextId() => Interlocked.Increment(ref _reqId);

    private HttpClient CreateClient()
    {
        var client = _httpFactory.CreateClient("Mcp");
        client.DefaultRequestHeaders.Clear();
        client.DefaultRequestHeaders.Add("x-api-key", _settings.ApiKey);
        client.DefaultRequestHeaders.Add("Accept", "application/json, text/event-stream");
        if (_sessionId != null)
            client.DefaultRequestHeaders.Add("mcp-session-id", _sessionId);
        return client;
    }

    private async Task EnsureSessionAsync()
    {
        if (_sessionId != null) return;
        await _sessionLock.WaitAsync();
        try
        {
            if (_sessionId != null) return;
            var client = CreateClient();

            var initBody = JsonSerializer.Serialize(new
            {
                jsonrpc = "2.0",
                method = "initialize",
                @params = new
                {
                    protocolVersion = "2024-11-05",
                    capabilities = new { },
                    clientInfo = new { name = "dashboard-dotnet", version = "1.0" }
                },
                id = NextId()
            });
            var initResp = await client.PostAsync(_settings.Url,
                new StringContent(initBody, Encoding.UTF8, "application/json"));

            if (initResp.Headers.TryGetValues("mcp-session-id", out var vals))
                _sessionId = vals.First();
            else if (initResp.Headers.TryGetValues("Mcp-Session", out var vals2))
                _sessionId = vals2.First();

            client = CreateClient();
            var notifBody = JsonSerializer.Serialize(new
            {
                jsonrpc = "2.0",
                method = "notifications/initialized"
            });
            await client.PostAsync(_settings.Url,
                new StringContent(notifBody, Encoding.UTF8, "application/json"));
        }
        finally
        {
            _sessionLock.Release();
        }
    }

    private void ResetSession() => _sessionId = null;

    public async Task<List<Dictionary<string, JsonElement>>> QueryAsync(
        string sql, string desc = "dashboard", string db = "InsappCoreProd")
    {
        await EnsureSessionAsync();

        for (int attempt = 0; attempt < 3; attempt++)
        {
            try
            {
                var client = CreateClient();
                var body = JsonSerializer.Serialize(new
                {
                    jsonrpc = "2.0",
                    method = "tools/call",
                    @params = new
                    {
                        name = "query",
                        arguments = new
                        {
                            database = db,
                            sql,
                            user_prompt = "dashboard",
                            query_description = desc
                        }
                    },
                    id = NextId()
                });

                using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(120));
                var resp = await client.PostAsync(_settings.Url,
                    new StringContent(body, Encoding.UTF8, "application/json"), cts.Token);

                var status = (int)resp.StatusCode;
                if (status is 401 or 403 or 404 or 410)
                {
                    ResetSession();
                    await EnsureSessionAsync();
                    if (attempt < 2) continue;
                    return [];
                }

                var text = await resp.Content.ReadAsStringAsync(cts.Token);
                if (string.IsNullOrWhiteSpace(text))
                {
                    if (attempt < 2) { ResetSession(); await EnsureSessionAsync(); continue; }
                    return [];
                }

                var ct = resp.Content.Headers.ContentType?.MediaType ?? "";
                if (ct == "text/event-stream")
                {
                    List<Dictionary<string, JsonElement>>? lastResult = null;
                    foreach (var line in text.Split('\n', '\r'))
                    {
                        if (!line.StartsWith("data: ")) continue;
                        try
                        {
                            var doc = JsonDocument.Parse(line[6..]);
                            var parsed = Parse(doc);
                            if (parsed != null) lastResult = parsed;
                        }
                        catch { }
                    }
                    if (lastResult != null) return lastResult;
                }
                else
                {
                    try
                    {
                        var doc = JsonDocument.Parse(text);
                        var parsed = Parse(doc);
                        if (parsed != null) return parsed;
                    }
                    catch { }
                }

                if (attempt < 2) { ResetSession(); await EnsureSessionAsync(); }
            }
            catch (Exception ex) when (ex is TaskCanceledException or HttpRequestException)
            {
                if (attempt < 2) { ResetSession(); await EnsureSessionAsync(); continue; }
            }
        }
        return [];
    }

    private static List<Dictionary<string, JsonElement>>? Parse(JsonDocument doc)
    {
        try
        {
            var root = doc.RootElement;
            if (!root.TryGetProperty("result", out var result)) return null;
            if (!result.TryGetProperty("content", out var content)) return null;

            foreach (var item in content.EnumerateArray())
            {
                if (item.TryGetProperty("type", out var t) && t.GetString() == "text" &&
                    item.TryGetProperty("text", out var textEl))
                {
                    var inner = JsonDocument.Parse(textEl.GetString()!);
                    var innerRoot = inner.RootElement;

                    if (innerRoot.TryGetProperty("error", out var err) &&
                        err.ValueKind == JsonValueKind.String &&
                        !string.IsNullOrEmpty(err.GetString()))
                        return [];

                    if (innerRoot.TryGetProperty("success", out var succ) && succ.GetBoolean() &&
                        innerRoot.TryGetProperty("rows", out var rows))
                    {
                        var list = new List<Dictionary<string, JsonElement>>();
                        foreach (var row in rows.EnumerateArray())
                        {
                            var dict = new Dictionary<string, JsonElement>();
                            foreach (var prop in row.EnumerateObject())
                                dict[prop.Name] = prop.Value.Clone();
                            list.Add(dict);
                        }
                        return list;
                    }
                }
            }
        }
        catch { }
        return null;
    }
}
